metrics/ci_integrations/test/client/buildkite/test_utils/mock/buildkite_mock_server.dart

```diff
@@ -0,0 +1,241 @@
+import 'dart:convert';
+import 'dart:io';
+import 'dart:math';
+import 'dart:typed_data';
+
+import 'package:api_mock_server/api_mock_server.dart';
+import 'package:ci_integration/client/buildkite/mappers/buildkite_build_state_mapper.dart';
+import 'package:ci_integration/client/buildkite/models/buildkite_artifact.dart';
+import 'package:ci_integration/client/buildkite/models/buildkite_build.dart';
+import 'package:ci_integration/client/buildkite/models/buildkite_build_state.dart';
+
+/// A mock server for the Buildkite API.
+class BuildkiteMockServer extends ApiMockServer {
+  /// A path to emulate a download url.
+  static const String _downloadPath = '/download';
+
+  /// Returns a base path of the Buildkite API.
+  String get basePath => '/organizations/organization_slug';
+
+  @override
+  List<RequestHandler> get handlers => [
+        RequestHandler.get(
+          pathMatcher: ExactPathMatcher(
+            '$basePath/pipelines/pipeline_slug/builds',
+          ),
+          dispatcher: _pipelineBuildsResponse,
+        ),
+        RequestHandler.get(
+          pathMatcher: ExactPathMatcher(
+            '$basePath/pipelines/not_found/builds',
+          ),
+          dispatcher: _notFoundResponse,
+        ),
+        RequestHandler.get(
+          pathMatcher: ExactPathMatcher(
+            '$basePath/pipelines/pipeline_slug/builds/1/artifacts',
+          ),
+          dispatcher: _pipelineBuildArtifactsResponse,
+        ),
+        RequestHandler.get(
+          pathMatcher: ExactPathMatcher(
+            '$basePath/pipelines/pipeline_slug/builds/not_found/artifacts',
+          ),
+          dispatcher: _notFoundResponse,
+        ),
+        RequestHandler.get(
+          pathMatcher: ExactPathMatcher(
+            '$basePath/pipelines/pipeline_slug/builds/1/artifacts/artifact_id/download',
+          ),
+          dispatcher: _downloadArtifactResponse,
+        ),
+        RequestHandler.get(
+          pathMatcher: ExactPathMatcher(
+            '$basePath/pipelines/pipeline_slug/builds/1/artifacts/not_found/download',
+          ),
+          dispatcher: _notFoundResponse,
+        ),
+        RequestHandler.get(
+          pathMatcher: ExactPathMatcher(_downloadPath),
+          dispatcher: _downloadResponse,
+        ),
+      ];
+
+  @override
+  List<AuthCredentials> get authCredentials => const [
+        AuthCredentials(token: 'token'),
+      ];
+
+  /// Responses with a list of all builds for a specific pipeline.
```

Comment:
Would it be great to specify that this method returns the builds with the `state` and `perPage` specified in the request? The same for all methods below.